examples/mem-alloc-test/src/main.rs

```diff
@@ -0,0 +1,275 @@
+ * per thread type i.e, if there are 2 small threads, 3 medium threads and
+ * 4 large threads each of which runs 10, 20 and 30 times, then
+ * avg_duration_small_thread will store average latency of 2*10=20 iterations,
+ * avg_duration_medium_thread will store average latency of 3*20=60 iterations,
+ * and avg_duration_large_thread will store average latency of 4*30=120 iterations
+ */
+struct Counters {
+    avg_duration_small_thread: f64,
+    avg_duration_medium_thread: f64,
+    avg_duration_large_thread: f64,
+    global_average: f64
+}
+
+const PAGE_SIZE: usize = 4096;
+const TO_KB: usize = 1024;
+const TO_MB: usize = TO_KB * 1024;
+const TO_GB: usize = TO_MB * 1024;
+
+const NUM_THREADS: i32 = 4;
+
+/* Set of configurable parameters. These will adjusted as necessary while
+ * recording the performance numbers
+ */
+
+const LIMIT_SMALL_THREAD: i32 = 2;
+const LIMIT_MEDIUM_THREAD: i32 = 2;
+const LIMIT_LARGE_THREAD: i32 = 2;
+
+const SCAN_INTERVAL_SMALL_THREAD: usize = 1 * TO_KB;
+const SCAN_INTERVAL_MEDIUM_THREAD: usize = 1 * TO_MB;
+const SCAN_INTERVAL_LARGE_THREAD: usize = PAGE_SIZE;
+
+
+const SMALL_THREAD_MEM_START:usize = 1;
+const SMALL_THREAD_MEM_END:usize = 512;
+const MEDIUM_THREAD_MEM_START:usize = 1;
+const MEDIUM_THREAD_MEM_END:usize = 128;
+const LARGE_THREAD_MEM_START:usize = 1;
+const LARGE_THREAD_MEM_END:usize = 4;
+
+
+
+fn calculate_and_print_stat(shared_mutex_clone: Arc<Mutex<Counters>>,
+                            tid: i32, memsize: &MemSize,
+                            avg_thread_latency: f64) {
+
+    /* TODO: Record some other statistical parameters like more detailed statistics
+     * than just average, such as standard deviation, minimum and maximum time,
+     * and p95/p99/p99.9 latency
+     */
+
+    let mut data = shared_mutex_clone.lock().unwrap();
+    /* Please note this is an intermediate value. Once we get the sum of individual
+     * averages of all the threads, then we will divide it by the frequency of
+     * the corresponding thread memsize type.
+     */
+    match memsize {
+        MemSize::Large => {
+            data.avg_duration_large_thread += avg_thread_latency;
+        }
+        MemSize::Medium => {
+            data.avg_duration_medium_thread += avg_thread_latency;
+        }
+        MemSize::Small => {
+            data.avg_duration_small_thread += avg_thread_latency;
+        }
+    };
+
+}
+
+fn get_random_num(start: usize, end: usize) -> usize {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(start..=end)
+}
+
+fn wait_per_thread(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
+    let (lock, cvar) = &*pair_clone;
+    let mut started = lock.lock().unwrap();
+    while !*started {
+        started = cvar.wait(started).unwrap();
+    }
+    drop(started);
+}
+
+fn wakeup_all_child_threads(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
+    let (lock, cvar) = &*pair_clone;
+    let mut started = lock.lock().unwrap();
+    *started = true;
+    drop(started);
```

Comment:
I don't think it's wrong to drop the lock before notifying the condition variable, but I think it's safe to do the notify with the lock still held and then just let the lock be dropped when the function returns.